### PR TITLE
fix panel rendering by transforming shortcuts data representation fro…

### PR DESCRIPTION
**Pull request by calvinnhieu**
…m DS to MDS, fix some callback hell bugs in background.js

- use let
- `getShortcutData` always invokes callback (typically render panel)
- `getPlugin` takes as param a `success` and a `failure` handler. The former invokes `mergeData`, the latter does not
- `mergeData` simply appends a Plugin section to [macro-data] shortcuts object instead of appending a plugin shortcut to its respective section since the plugin system doesn't currently support specifying sections per shortcut

Note: Don't think we need to be writing so functionally - our program's pretty straightforward so I don't think there's much value in reusability and tracing these callbacks is kind of a pain. I think the code could be cleaned up if we didn't rely on callbacks as much, but I'm not sure it's 3AM lol.

### CODE REVIEW

File: chrome-extension/js/background.js [hunk 1]
```diff
@@ -69,75 +69,105 @@ function extractRootDomain(url) {
...
+  let xhr = new XMLHttpRequest();
   xhr.open('GET', key, true);
   xhr.onload = (e) => {
     if (xhr.readyState === 4) {
       if (xhr.status === 200) {
-        var json = JSON.parse(xhr.responseText);
+        let json = JSON.parse(xhr.responseText);
         callback(json);
       } else {
+        callback({sections: []});
```
Comment (calvinnhieu) [inline chrome-extension/js/background.js]: dummy MD (macro-data) shortcuts object

File: chrome-extension/js/background.js [hunk 2]
```diff
@@ -190,50 +220,51 @@ chrome.tabs.onUpdated.addListener((tabId, changeInfo, tab) => {
...
-        "macos": ["k"]
-      }],
-      action: () => {
-        alert('New section plugin');
-      }
-    }
-  ];
+// currently supports 1 plugin per domain
+function getPlugin(domain, success, failure) {
+  domain in Plugins ? success(Plugins[domain]) : failure();
```
Comment (calvinnhieu) [inline chrome-extension/js/background.js:203]: sufficient (MVP)?

File: chrome-extension/plugin/messenger.ts [hunk 3]
```diff
@@ -22,7 +22,7 @@ class MessengerPage {
   }
 
   getUnreadLinks() : HTMLAnchorElement[] {
-    return document.querySelectorAll('li[aria-relevant="additions text"][aria-live="polite"]');
+    return Array.prototype.slice.call(document.querySelectorAll('li[aria-relevant="additions text"][aria-live="polite"]'));
```
Comment (calvinnhieu) [inline chrome-extension/plugin/messenger.ts:25]: `querySelectorAll` returns a NodeList, not an Array.
https://hackernoon.com/htmlcollection-nodelist-and-array-of-objects-da42737181f9

File: chrome-extension/plugin/pluginbuilder.ts [hunk 4]
```diff
@@ -22,7 +22,7 @@ class PluginState {
 
 export class Plugin {
   domain: string;
-  shortcuts: any[];
+  shortcuts: any[]; // MDS
```
Comment (calvinnhieu) [inline chrome-extension/plugin/pluginbuilder.ts]: re: shortcuts data representation
MDS = macro-data schema (macro-data, rendering panel)
DS = developer schema (in Plugins)

File: chrome-extension/js/background.js [hunk 5]
```diff
@@ -69,75 +69,105 @@ function extractRootDomain(url) {
...
+  let pluginSection = {name: 'Plugins', description: 'Shortcuts from plugins', shortcuts: []}
+  pluginSection.shortcuts = plugins.map(plugin => {
+    return {
+      name: plugin.name,
+      keys: plugin.keys
+    };
+  });
+  shortcuts.sections.push(pluginSection);
   return shortcuts;
+
```
Comment (andyzg) [inline chrome-extension/js/background.js]: can cut all of the below, it's pretty confusing what's going on.

File: chrome-extension/js/background.js [hunk 6]
```diff
@@ -69,75 +69,105 @@ function extractRootDomain(url) {
...
+    (plugin) => {
+      let domainKey = getDomainKey(url);
+      getShortcutData(domainKey, (shortcuts) => {
+        let data = mergeData(shortcuts, plugin.default.shortcuts);
+        save(domainKey, data);
+        render(data);
+      });
+      initPlugin(plugin);
+    },
+    // failure handler: if no plugins, use MD shortcuts, cache & render
```
Comment (andyzg) [inline chrome-extension/js/background.js]: MD?

File: chrome-extension/js/background.js [hunk 7]
```diff
@@ -154,11 +184,11 @@ function isEmpty(obj) {
 chrome.commands.onCommand.addListener((command) => {
   if (command == 'toggle-popup') {
     getCurrentTabUrl((url) => {
-      var key = getKey(url)
+      let key = getDomainKey(url);
```
Comment (andyzg) [inline chrome-extension/js/background.js]: should we load the script on the frontend on load, and then check for the command pop-up on the client side? That way, we don't have to continuously message back n forth?
Can do for another PR

File: chrome-extension/js/background.js (hunk 2 above)
Comment (andyzg) [reply chrome-extension/js/background.js:203]: i dig it

File: chrome-extension/plugin/messenger.ts (hunk 3 above)
Comment (andyzg) [reply chrome-extension/plugin/messenger.ts:25]: oh yeah, good call, thanks for fixing this

File: chrome-extension/plugin/pluginbuilder.ts (hunk 4 above)
Comment (andyzg) [reply chrome-extension/plugin/pluginbuilder.ts]: ouh, gotcha

**Review by andyzg — APPROVED** [review]
sexy 1.0

File: chrome-extension/js/background.js [hunk 8]
```diff
@@ -69,75 +69,84 @@ function extractRootDomain(url) {
...
+    name: 'Plugins',
+    description: 'Shortcuts from plugins',
+    shortcuts: plugins
+  };
+  shortcuts.sections.push(pluginSection);
   return shortcuts;
 }
 
-function initPlugins(plugins) {
+// TODO: test
```
Comment (chrisjluc) [inline chrome-extension/js/background.js]: is this tested?

**Review by chrisjluc — APPROVED** [review]

File: chrome-extension/js/background.js (hunk 8 above)
Comment (calvinnhieu) [reply chrome-extension/js/background.js]: Oops it is! removing todo

File: chrome-extension/js/background.js (hunk 7 above)
Comment (calvinnhieu) [reply chrome-extension/js/background.js]: Don't think so. I think the command pop-up listener should be the chrome extension's responsibility. Plus if we move this hook to the client, it looks like we'd have to move the bulk of the logic/fetching data to the client as well. Thoughts?
